File: example/node/lib/rest.js
const http = require('http');
var utils = require('./utils');

function httpApi(params, callback){
  const options = {
    protocol: "http:", hostname: "localhost", port: process.env.NT_HTTP_PORT,
    path: "/api/v6"+params.path, method: params.method,
    headers: {
      "Content-Type": "application/json",
    }
  };
  const post_data = (params.data) ? utils.EncodeOptions(params.data) : null
  if(params.token){
    options.headers["Authorization"] = "Bearer "+params.token
  }
  if(process.env.NT_API_KEY){
    options.headers["X-API-Key"] = process.env.NT_API_KEY
  }
  if(post_data){
    options.headers["Content-Length"] = post_data.length
  }
  const hreq = http.request(options, (hres) => {
    let data = "";

    hres.on("data", (chunk) => {
        data += chunk;
    });

    hres.on("end", () => {
      callback(utils.CheckJson(data))
    });

  }).on("error", (err) => {
    callback({code:400, message: err.message})
  });
  
  if(post_data){
    hreq.write(post_data);
  }
  hreq.end();
}

exports.Get = function(token, path, query, callback) {
  var path = "/"+path;
  var query_str = "";
  for(var key in query){
    query_str += "&"+key+"="+query[key];
  }
  if(query_str){
    path += "?"+encodeURIComponent(query_str);
  }
  httpApi({ path, method: "GET", token }, function(data){
    callback(data)
  })
}

exports.Post = function(token, path, post_data, callback) {
  var path = "/"+path;
  httpApi({ path, method: "POST", token, data: post_data }, function(data){
    callback(data)
  })
}

exports.Put = function(token, path, post_data, callback) {
  var path = "/"+path;
  httpApi({ path, method: "PUT", token, data: post_data }, function(data){
    callback(data)
  })
}

exports.Delete = function(token, path, callback) {
  var path = "/"+path;
  httpApi({ path, method: "DELETE", token }, function(data){
    callback(data)
  })
}

File: example/go/service/rest.go
package service

import (
	"bytes"
	"errors"
	"io"
	"net/http"
	"net/url"
	"slices"

	cu "github.com/nervatura/component/pkg/util"
)

type HttpClient struct {
	Config cu.SM
}

func (rest *HttpClient) request(method, path, token string, options cu.IM) (any, error) {
	service_url := "http://localhost:" + rest.Config["NT_HTTP_PORT"] + "/api/v6/"
	data, err := cu.ConvertToByte(options)
	if err != nil {
		return nil, err
	}

	request, err := http.NewRequest(method, service_url+path, bytes.NewBuffer(data))
	if err != nil {
		return nil, err
	}
	request.Header.Set("Content-Type", "application/json; charset=UTF-8")
	if token != "" {
		request.Header.Set("Authorization", "Bearer "+token)
	} else {
		request.Header.Set("X-API-Key", rest.Config["NT_API_KEY"])
	}

	client := &http.Client{}
	response, err := client.Do(request)
	if err != nil {
		return nil, err
	}
	defer response.Body.Close()

	if response.Header.Get("Content-Type") == "application/pdf" {
		responseStr := ""
		responseData, err := io.ReadAll(response.Body)
		if err == nil {
			responseStr = string(responseData)
		}
		return responseStr, err
	}

	var result interface{}
	if response.StatusCode == 200 {
		err = cu.ConvertFromReader(response.Body, &result)
		if err != nil {
			return nil, errors.New(response.Status)
		}
	}
	if !slices.Contains([]int64{200, 201, 204}, int64(response.StatusCode)) {
		return nil, errors.New(response.Status)
	}
	return result, err
}

func (rest *HttpClient) Get(token, path string, query cu.IM) (any, error) {
	queryParams := url.Values{}
	for key, value := range query {
		queryParams.Add(key, cu.ToString(value, ""))
	}
	if len(queryParams) > 0 {
		path += "?" + queryParams.Encode()
	}
	return rest.request("GET", path, token, cu.IM{})
}

func (rest *HttpClient) Post(token, path string, data cu.IM) (any, error) {
	return rest.request("POST", path, token, data)
}

func (rest *HttpClient) Put(token, path string, data cu.IM) (any, error) {
	return rest.request("PUT", path, token, data)
}

func (rest *HttpClient) Delete(token, path string) (any, error) {
	return rest.request("DELETE", path, token, cu.IM{})
}
